buffer/src/main/java/io/netty/buffer/EmptyByteBuf.java

```diff
@@ -105,6 +105,12 @@ public boolean isReadOnly() {
         return false;
     }
 
+    /**
+     * An empty buffer is a special buffer, it is both heap and direct,
+     * which means if there is a isHeap() method in this class, it will return {@code true} too.
+     * See <a href="https://github.com/netty/netty/issues/11354#issuecomment-855193484">this</a>
+     * for detail explanation.
+     */
```

Comment:
I guess most users will not take a look into the Javadoc of `EmptyByteBuf` because it's not part of public API. How about just adding a couple sentences like the following to `Unpooled.EMPTY_BUFFER`?

```java
    /**
     * A empty buffer whose capacity is {@code 0}. Note that an empty buffer is both direct
     * and heap buffer, because it is always backed by both a zero-length byte array and
     * direct buffer internally.
     */
    public static final ByteBuf EMPTY_BUFFER = ALLOC.buffer(0, 0);
```

Reply:
@trustin , The reason I added the doc to `EmptyByteBuf.isDirect()` is: The doc of its parent method `ByteBuf.isDirect()` states: 
>   * Returns {@code true} if and only if this buffer is backed by an
>   * NIO direct buffer.

Note the 'if and only if this buffer is backed by an NIO direct buffer' which is not a consistent logic with `EmptyByteBuf.isDirect()`, so I think it needs the additional docs to explain it.

Reply:
I see. Then how about updating the Javadoc of `ByteBuf.isDirect()`?

Reply:
I think it is OK to update the doc of `ByteBuf`, then the following methods's doc from `ByteBuf` may also need to be updated:

> ByteBuf.hasArray();
> ByteBuf.hasMemoryAddress();

because they all have doc states like "if and only if". WDYT?

Reply:
SGTM. Please go ahead.

Reply:
/cc @chrisvest You might want to update the Javadoc of the new API in your PR likewise.

Reply:
@trustin I think behaviour should be consistent as much as possible. In the new API the only thing special about an empty buffer is that it can't be directly allocated, but it can be created in other ways.

Reply:
@chrisvest Yeah, what I'm suggesting is to mention that an empty buffer has both array and memory address unlike other buffers. Do we already?

Reply:
docs updated.

Reply:
@trustin In the new API even empty buffers are either on-heap or off-heap, not some mix. There's nothing special about empty buffers in the new API, so there's nothing to document.

Reply:
@chrisvest I see. Thanks for letting me know!